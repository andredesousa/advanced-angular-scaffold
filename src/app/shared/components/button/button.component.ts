import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';

@Component({
  selector: 'app-button',
  template: ` <button
    type="button"
    [ngClass]="classes"
    [ngStyle]="{ 'background-color': backgroundColor }"
    (click)="clicked.emit($event)"
  >
    {{ label }}
  </button>`,
  styleUrls: ['./button.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ButtonComponent {
  /**
   * Is this the principal call to action on the page?
   */
  @Input()
  primary = false;

  /**
   * What background color to use
   */
  @Input()
  backgroundColor?: string;

  /**
   * How large should the button be?
   */
  @Input()
  size: 'small' | 'medium' | 'large' = 'medium';

  /**
   * Button contents
   *
   * @required
   */
  @Input()
  label = 'Button';

  /**
   * Optional click handler
   */
  @Output()
  clicked = new EventEmitter<Event>();

  get classes(): string[] {
    const mode = this.primary ? 'app-button--primary' : 'app-button--secondary';

    return ['app-button', `app-button--${this.size}`, mode];
  }
}
