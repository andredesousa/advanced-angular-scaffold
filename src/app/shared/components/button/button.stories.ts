import { Story, Meta } from '@storybook/angular';
import { action } from '@storybook/addon-actions';

import { ButtonComponent } from './button.component';

export default {
  title: 'Example/Button',
  component: ButtonComponent,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta;

const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
  component: ButtonComponent,
  props: args,
});

export const Primary = Template.bind({});

Primary.args = {
  primary: true,
  label: 'Button',
  clicked: action('clicked'),
};

export const Secondary = Template.bind({});

Secondary.args = {
  label: 'Button',
  clicked: action('clicked'),
};

export const Large = Template.bind({});

Large.args = {
  size: 'large',
  label: 'Button',
  clicked: action('clicked'),
};

export const Small = Template.bind({});

Small.args = {
  size: 'small',
  label: 'Button',
  clicked: action('clicked'),
};
