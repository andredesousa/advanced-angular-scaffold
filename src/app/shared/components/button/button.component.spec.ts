import { TestBed } from '@angular/core/testing';
import { ButtonComponent } from './button.component';

describe('ButtonComponent', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ButtonComponent],
    }).compileComponents();
  });

  it('should return an array of CSS classes with "app-button--primary"', () => {
    const fixture = TestBed.createComponent(ButtonComponent);
    const component = fixture.componentInstance;

    component.primary = true;

    expect(component.classes).toEqual(['app-button', 'app-button--medium', 'app-button--primary']);
  });

  it('should return an array of CSS classes with "app-button--secondary"', () => {
    const fixture = TestBed.createComponent(ButtonComponent);
    const component = fixture.componentInstance;

    expect(component.classes).toEqual(['app-button', 'app-button--medium', 'app-button--secondary']);
  });

  it('should match with the snapshot', () => {
    const fixture = TestBed.createComponent(ButtonComponent);

    expect(fixture).toMatchSnapshot();
  });
});
