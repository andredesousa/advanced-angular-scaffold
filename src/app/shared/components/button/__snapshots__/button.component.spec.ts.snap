// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent should match with the snapshot 1`] = `
<app-button
  __ngContext__={[Function LRootView]}
  clicked={[Function EventEmitter_]}
  label={[Function String]}
  primary="false"
  size={[Function String]}
>
  <button
    type="button"
  >
    
  </button>
</app-button>
`;
